pkg/common/peertracker/peertracker_test_child_windows.go

```diff
@@ -0,0 +1,195 @@
+//go:build ignore
+// +build ignore
+
+// This file is used during testing. It is built as an external binary
+// and called from the test suite in order to exercise various peer
+// tracking scenarios
+package main
+
+import (
+	"errors"
+	"flag"
+	"fmt"
+	"net"
+	"os"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+var (
+	modWs2_32 = windows.NewLazySystemDLL("ws2_32.dll")
+
+	// WSADuplicateSocketA function (winsock2.h)
+	procWSADuplicateSocketA = modWs2_32.NewProc("WSADuplicateSocketA")
+)
+
+func main() {
+	var (
+		tcpSocketPort    int
+		protocolInfoFile string
+	)
+
+	flag.IntVar(&tcpSocketPort, "tcpSocketPort", -1, "port to peertracker tcp socket")
+	flag.StringVar(&protocolInfoFile, "protocolInfoFile", "", "path to file containing WSAPROTOCOL_INFO structure")
+	flag.Parse()
+
+	if protocolInfoFile == "" {
+		fmt.Fprint(os.Stderr, "-protocolInfoFile flag is required")
+		os.Exit(1)
+	}
+
+	// We are a grandchild - send a sign then sleep forever
+	if tcpSocketPort == -1 {
+		piBytes, err := getProtocolInfo(protocolInfoFile)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error reading file with WSAPROTOCOL_INFO structure: %v", err)
+			os.Exit(2)
+		}
+
+		pi := (*windows.WSAProtocolInfo)(unsafe.Pointer(&piBytes[0]))
+		fd, err := windows.WSASocket(-1, -1, -1, pi, 0, 0)
+
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "WSASocket error: %v", err)
+			os.Exit(3)
+		}
+
+		err = send(fd, "i'm alive!")
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Send failed: %v", err)
+			os.Exit(4)
+		}
+
+		// sleep forever
+		select {}
+	}
+
+	addr := &net.TCPAddr{
+		IP:   net.IPv4(127, 0, 0, 1),
+		Port: tcpSocketPort,
+	}
+
+	conn, err := net.DialTCP("tcp", nil, addr)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "DialTCP failed: %v", err)
+		os.Exit(5)
+	}
+
+	socket, err := getFd(conn)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to get file descriptor from connection: %v", err)
+		os.Exit(6)
+	}
+
+	procattr := &syscall.ProcAttr{
+		Env:   os.Environ(),
+		Files: []uintptr{0, 0, 0}, // Do not block on stdin / stdout / stderr
+	}
+
+	pid, _, err := syscall.StartProcess(os.Args[0], []string{os.Args[0], "-protocolInfoFile", protocolInfoFile}, procattr)
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to produce grandchild: %v", err)
+		os.Exit(7)
+	}
+
+	if err := duplicateSocket(windows.Handle(socket), pid, protocolInfoFile); err != nil {
```

Comment:
Why do we duplicate this after the grandchild has started?

Reply:
Because we need to know the PID of the grandchild to call the WSADuplicateSocket function. Please refer to the sample scenario here:
https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsaduplicatesocketa#remarks

Reply:
Ah, missed that. Thanks :)